metronome/checkpointing/service/test/src/io/iohk/metronome/checkpointing/service/InterpreterClientSpec.scala

```diff
@@ -0,0 +1,333 @@
+package io.iohk.metronome.checkpointing.service
+
+import cats.implicits._
+import cats.effect.Resource
+import cats.effect.concurrent.Ref
+import io.iohk.metronome.checkpointing.models.{
+  Transaction,
+  Ledger,
+  Block,
+  CheckpointCertificate
+}
+import io.iohk.metronome.checkpointing.models.ArbitraryInstances
+import io.iohk.metronome.checkpointing.CheckpointingAgreement
+import io.iohk.metronome.checkpointing.interpreter.ServiceRPC
+import io.iohk.metronome.checkpointing.interpreter.messages.InterpreterMessage
+import io.iohk.metronome.checkpointing.interpreter.InterpreterRPC
+import io.iohk.metronome.checkpointing.service.tracing.CheckpointingEvent
+import io.iohk.metronome.networking.LocalConnectionManager
+import io.iohk.metronome.tracer.Tracer
+import monix.catnap.ConcurrentQueue
+import monix.eval.Task
+import monix.execution.schedulers.TestScheduler
+import monix.tail.Iterant
+import org.scalatest.flatspec.AnyFlatSpec
+import org.scalatest.matchers.should.Matchers
+import org.scalatest.compatible.Assertion
+import org.scalatest.Inspectors
+import org.scalacheck.Arbitrary
+import scala.concurrent.duration._
+
+class InterpreterClientSpec extends AnyFlatSpec with Matchers with Inspectors {
+  import InterpreterClientSpec._
+  import InterpreterMessage._
+  import ArbitraryInstances._
+
+  def sample[T: Arbitrary] =
+    implicitly[Arbitrary[T]].arbitrary.sample.get
+
+  def test(fixture: Fixture): Assertion = {
+    implicit val scheduler = TestScheduler()
+
+    val fut =
+      fixture.resources.use(fixture.test).delayExecution(0.second).runToFuture
+
+    // These tests shouldn't take too long, so just simulate some generous amount of time.
+    scheduler.tick(1.minute)
```

Comment:
I'm curious why this (and `delayExecution`) is needed

Reply:
Excellent question. I'm sorry, I left a [comment](https://github.com/input-output-hk/metronome/pull/44/files#diff-31f9fae33dbc8e276e1061b50f1d73e06c1f904ab986f8948509de386f4fda76R33) in the `RobotIntegrationSpec`. 

When I first used the `TestScheduler` I noticed that some of the `Task`s I schedule get executed when I call `tick()` but others run earlier. The reason was that if a task has no delay, it will be just executed right there, so the `TestScheduler` can work through everything a delayed task triggered until it hits something that is scheduled for the future. 

Not sure why it was a problem in those tests I was working on, but I thought it would be easier to understand things if I set up the test and know that it will only do something when I pull the trigger. Adding even a 0 delay means nothing happens until this line.

Here's a demonstration:

```console
$ mill -i metronome[2.13.4].core.console
[31/31] metronome[2.13.4].core.console 
Welcome to Scala 2.13.4 (OpenJDK 64-Bit Server VM, Java 11.0.11).
Type in expressions for evaluation. Or try :help.

scala> import monix.execution.schedulers.TestScheduler
import monix.execution.schedulers.TestScheduler

scala> import monix.eval.Task
import monix.eval.Task

scala> val t = Task(println("Hello"))
val t: monix.eval.Task[Unit] = Task.Eval$1605702850

scala> implicit val scheduler = TestScheduler()
val scheduler: monix.execution.schedulers.TestScheduler = monix.execution.schedulers.TestScheduler@215c6ec0

scala> val f = t.runToFuture
Hello
val f: monix.execution.CancelableFuture[Unit] = monix.execution.CancelableFuture$Pure@3ba8ef4e

scala> import scala.concurrent.duration._
import scala.concurrent.duration._

scala> val t0 = Task(println("World")).delayExecution(Duration.Zero)
val t0: monix.eval.Task[Unit] = Task.FlatMap$719355325

scala> val f0 = t0.runToFuture
val f0: monix.execution.CancelableFuture[Unit] = Async(Future(<not completed>),monix.eval.internal.TaskConnection$Impl$$anon$1@3100b5cf,null)

scala> scheduler.tick()
World
```

As to why `scheduler.tick(1.minute)` is needed: there are timeouts involved in some of the tests, for example [this one](https://github.com/input-output-hk/metronome/pull/53/files/938f5b56ac79bce6b7ffb8e180411a38cb476e55#diff-080bb23b02bac10ccb5fd2b77c71f79c6f5d8fcf52c26543902b0fb9ecc90c97R98). These tests all use mocks, not a real network (unlike the ones in https://github.com/input-output-hk/metronome/pull/51 which start an actual Scalanet peer groups), so we don't actually have to wait out the 5 second timeout, or set them to reasonably low values; we can use the `TestScheduler` to simulate the passage of time. 

In the `RobotIntegrationSpec` I added a [duration property](https://github.com/input-output-hk/metronome/pull/44/files#diff-31f9fae33dbc8e276e1061b50f1d73e06c1f904ab986f8948509de386f4fda76R175-R178) to the `Fixture` to say how long it's going to run for, but here they are all just "some time with a potential timeout" so 1 minute should be plenty. 

I don't want to just keep ticking until there are no more tasks because there could be something on a loop, so this gives a timeout for the test as well. It won't actually do anything if there are no tasks.

Reply:
It's nice in that when I call `enqueueMessage` it will add a message to a `ConcurrentQueue` which is consumed on a different `Fiber`, so I don't know exactly when I can expect the effects to be observable. I know that the execution is _almost_ instantaneous, but not quite.

I could maybe add some `Deferred` somewhere that I can wait upon, but it could be awkward to do. Or I can call `Task.sleep` in the test, but how long should I sleep for? What works locally always turns out to be too tight on CI.

With the `TestScheduler`, the non-delayed tasks will be executed first, so I can add something like this `awaitProcessing` with a sleep time that I can trust is surely enough for something that is processed immediately, and then all it will do is jump the clock by 5 seconds and carry on with the rest of my test, doing the assertions, without actually wasting any time.

Reply:
Interesting, thanks for the explanation